mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java

```diff
@@ -0,0 +1,173 @@
+  }
+
+  public void setHiveSessionValue(String key, String value) {
+    checkState(session != null, "There is no open session for setting variables.");
+    try {
+      session.getSessionConf().set(key, value);
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to set Hive session variable: ", e);
+    }
+  }
+
+  public void start() {
+    checkState(isNotBlank(conf.get(HiveConf.ConfVars.METASTOREURIS.varname)),
+            "hive.metastore.uris must be supplied in config. TestHiveShell needs an external metastore to connect to.");
+    hs2.init(conf);
+    hs2.start();
+    client = hs2.getServices().stream()
+            .filter(CLIService.class::isInstance)
+            .findFirst()
+            .map(CLIService.class::cast)
+            .get();
+    started = true;
+  }
+
+  public void stop() {
+    if (client != null) {
+      client.stop();
+    }
+    hs2.stop();
+    started = false;
+  }
+
+  public void openSession() {
+    checkState(started, "You have to start TestHiveShell first, before opening a session.");
+    try {
+      SessionHandle sessionHandle = client.getSessionManager().openSession(
+              CLIService.SERVER_VERSION, "", "", "127.0.0.1", Collections.emptyMap());
+      session = client.getSessionManager().getSession(sessionHandle);
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to open new Hive session: ", e);
+    }
+  }
+
+  public void closeSession() {
+    checkState(session != null, "There is no open session to be closed.");
+    try {
+      session.close();
+      session = null;
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to close Hive session: ", e);
+    }
+  }
+
+  public List<Object[]> executeStatement(String statement) {
+    checkState(session != null,
+            "You have to start TestHiveShell and open a session first, before running a query.");
+    try {
+      OperationHandle handle = client.executeStatement(session.getSessionHandle(), statement, Collections.emptyMap());
+      List<Object[]> resultSet = new ArrayList<>();
+      if (handle.hasResultSet()) {
+        RowSet rowSet;
+        // keep fetching results until we can
+        while ((rowSet = client.fetchResults(handle)) != null && rowSet.numRows() > 0) {
+          for (Object[] row : rowSet) {
+            resultSet.add(row.clone());
+          }
+        }
+      }
+      return resultSet;
+    } catch (HiveSQLException e) {
+      throw new IllegalArgumentException("Failed to execute Hive query '" + statement + "': " + e.getMessage(), e);
+    }
+  }
+
+  public Configuration getHiveConf() {
+    if (session != null) {
+      return session.getHiveConf();
+    } else {
+      return conf;
+    }
+  }
+
+  private HiveConf initializeConf() {
+    HiveConf hiveConf = new HiveConf();
+
+    // Use random port to enable running tests in parallel
+    hiveConf.setIntVar(HiveConf.ConfVars.HIVE_SERVER2_THRIFT_PORT, FIND_FREE_PORT.invoke());
```

Comment:
Isn't there an option to get an ephemeral port when creating the socket in the listener? Seems odd to find a free port and try to use it. There's a small chance of collision with new connections on the host.

Reply:
You're right, there is a chance of collision. I've changed the port number to 0 in the config, which will make HS2 use an ephemeral port.